include/boost/geometry/algorithms/difference.hpp

```diff
@@ -268,10 +299,291 @@ inline OutputIterator difference_insert(Geometry1 const& geometry1,
+struct multi_output_type<Geometry, Collection, linear_tag>
+{
+    using type = typename util::sequence_find_if
+        <
+            typename traits::geometry_types<Collection>::type,
+            util::is_multi_linestring
+        >::type;
+};
+
+template <typename Geometry, typename Collection>
+struct multi_output_type<Geometry, Collection, areal_tag>
+{
+    using type = typename util::sequence_find_if
+        <
+            typename traits::geometry_types<Collection>::type,
+            util::is_multi_polygon
+        >::type;
+};
+
+
 }} // namespace detail::difference
 #endif // DOXYGEN_NO_DETAIL
 
 
+namespace resolve_collection
+{
+
+
+template
+<
+    typename Geometry1, typename Geometry2, typename Collection,
+    typename Tag1 = typename geometry::tag<Geometry1>::type,
+    typename Tag2 = typename geometry::tag<Geometry2>::type,
+    typename CollectionTag = typename geometry::tag<Collection>::type
+>
+struct difference
+{
+    template <typename Strategy>
+    static void apply(Geometry1 const& geometry1,
+                      Geometry2 const& geometry2,
+                      Collection & output_collection,
+                      Strategy const& strategy)
+    {
+        using single_out = typename geometry::detail::output_geometry_value
+            <
+                Collection
+            >::type;
+
+        detail::difference::difference_insert<single_out>(
+            geometry1, geometry2,
+            geometry::detail::output_geometry_back_inserter(output_collection),
+            strategy);
+    }
+};
+
+template <typename Geometry1, typename Geometry2, typename Collection>
+struct difference
+    <
+        Geometry1, Geometry2, Collection,
+        geometry_collection_tag, geometry_collection_tag, geometry_collection_tag
+    >
+{
+    template <typename Strategy>
+    static void apply(Geometry1 const& geometry1,
+                      Geometry2 const& geometry2,
+                      Collection& output_collection,
+                      Strategy const& strategy)
+    {
+        auto const rtree2 = detail::make_rtree_iterators(geometry2, strategy);
+        detail::visit_breadth_first([&](auto const& g1)
+        {
+            // multi-point, multi-linestring or multi_polygon
+            typename detail::difference::multi_output_type
+                <
+                    util::remove_cref_t<decltype(g1)>, Collection
+                >::type out;
+
+            g1_minus_gc2(g1, rtree2, out, strategy);
+
+            if (! boost::empty(out))
+            {
+                if (boost::size(out) == 1)
+                {
+                    move_single_out(output_collection, out);
+                }
+                else
+                {
+                    range::emplace_back(output_collection, std::move(out));
+                }
+            }
+
+            return true;
+        }, geometry1);
+    }
+
+private:
+    // Implemented as separate function because msvc is unable to do nested lambda capture
```

Comment:
what version of MSVC does not support this?

Reply:
I tested with msvc-14.1 i.e. VS2017.

Reply:
OK thanks! Side note, it could have been fixed in VS2019 
https://developercommunity.visualstudio.com/t/this-captured-in-nested-lambda-when-used-in-captur/292128